crates/flipperzero/src/furi/string.rs

```diff
@@ -0,0 +1,259 @@
+//! String primitives built around `FuriString`.
+
+use core::{
+    cmp::Ordering,
+    convert::Infallible,
+    ffi::{c_char, CStr},
+    fmt, ops,
+};
+
+#[cfg(feature = "alloc")]
+use alloc::ffi::CString;
+
+use flipperzero_sys as sys;
+
+/// A Furi string.
+///
+/// This is similar to Rust's [`String`], but used inside the Flipper Zero SDK.
+#[derive(Eq)]
+pub struct String(*mut sys::FuriString);
```

Comment:
Looking at [`furi/core/string.h`](https://sourcegraph.com/github.com/flipperdevices/flipperzero-firmware/-/blob/furi/core/string.h) and the underlying [m-string](https://github.com/P-p-H-d/mlib/tree/master#m-string), I think we can safely define it as `String(sys::FuriString)` so we can take advantage of m-string's ability to store small strings entirely on the stack.

Reply:
I don't think that helps? [`sys::furi_string_alloc()` always allocates a `FuriString` on the heap](https://github.com/flipperdevices/flipperzero-firmware/blob/c1bb10a6947f56e7cb71912b864b6d2b969f6155/furi/core/string.c#L25-L29), regardless of how the `string_t` underneath is allocated.

The Flipper Zero SDK does not expose the `FuriString` C struct directly; the `sys::FuriString` type is being generated by `bindgen` as a target for the `*mut FuriString` pointer. If we define this as `String(sys::FuriString)` then we'd effectively only be able to represent `&mut String` and `&String`, not an owned type.

Reply:
That's correct. `sys::furi_string_alloc` always allocates a `FuriString` on the heap, but there's nothing stopping you from putting an uninitialized `FuriString` on the stack and then calling `string_init(&furi_string.string)` on that structure (other than there being no `sys::furi_string_init`—I'll see if they'll accept a PR).

Reply:
Oh, I see what you mean. It's an opaque type, so it's not possible to even know `sizeof(FuriString)`.

Still, I'll see if it might be possible to expose that detail.

Reply:
Filed a feature request on the SDK: https://github.com/flipperdevices/flipperzero-firmware/issues/2461

Reply:
In any case, we don't need to block on this, as `furi::String` is an opaque type, so we can change its internals whenever we want.

Reply:
I agree. Let's move forward with the current implementation.